File: src-tauri/src/commands/screenshot.rs
//! Screenshot Tauri commands
//!
//! Commands for capturing screenshots and performing OCR.

use crate::screenshot::{
    CaptureRegion, MonitorInfo, ScreenshotConfig, ScreenshotManager, ScreenshotMetadata,
    ScreenshotHistoryEntry, WinOcrResult, WindowInfo, SnapConfig, SnapResult,
};
use tauri::State;

/// Capture full screen screenshot
#[tauri::command]
pub async fn screenshot_capture_fullscreen(
    manager: State<'_, ScreenshotManager>,
    monitor_index: Option<usize>,
) -> Result<ScreenshotResult, String> {
    let result = manager.capture_fullscreen(monitor_index).await?;
    Ok(ScreenshotResult {
        image_base64: base64::engine::general_purpose::STANDARD.encode(&result.image_data),
        metadata: result.metadata,
    })
}

/// Capture active window screenshot
#[tauri::command]
pub async fn screenshot_capture_window(
    manager: State<'_, ScreenshotManager>,
) -> Result<ScreenshotResult, String> {
    let result = manager.capture_window().await?;
    Ok(ScreenshotResult {
        image_base64: base64::engine::general_purpose::STANDARD.encode(&result.image_data),
        metadata: result.metadata,
    })
}

/// Capture region screenshot
#[tauri::command]
pub async fn screenshot_capture_region(
    manager: State<'_, ScreenshotManager>,
    x: i32,
    y: i32,
    width: u32,
    height: u32,
) -> Result<ScreenshotResult, String> {
    let region = CaptureRegion { x, y, width, height };
    let result = manager.capture_region(region).await?;
    Ok(ScreenshotResult {
        image_base64: base64::engine::general_purpose::STANDARD.encode(&result.image_data),
        metadata: result.metadata,
    })
}

/// Start interactive region selection
#[tauri::command]
pub async fn screenshot_start_region_selection(
    manager: State<'_, ScreenshotManager>,
) -> Result<CaptureRegion, String> {
    manager.start_region_selection().await
}

/// Extract text from image using OCR
#[tauri::command]
pub async fn screenshot_ocr(
    manager: State<'_, ScreenshotManager>,
    image_base64: String,
) -> Result<String, String> {
    let image_data = base64::engine::general_purpose::STANDARD
        .decode(&image_base64)
        .map_err(|e| format!("Failed to decode image: {}", e))?;
    
    manager.extract_text(&image_data)
}

/// Get list of available monitors
#[tauri::command]
pub async fn screenshot_get_monitors(
    manager: State<'_, ScreenshotManager>,
) -> Result<Vec<MonitorInfo>, String> {
    Ok(manager.get_monitors())
}

/// Update screenshot configuration
#[tauri::command]
pub async fn screenshot_update_config(
    manager: State<'_, ScreenshotManager>,
    config: ScreenshotConfig,
) -> Result<(), String> {
    manager.update_config(config);
    Ok(())
}

/// Get current screenshot configuration
#[tauri::command]
pub async fn screenshot_get_config(
    manager: State<'_, ScreenshotManager>,
) -> Result<ScreenshotConfig, String> {
    Ok(manager.get_config())
}

/// Save screenshot to file
#[tauri::command]
pub async fn screenshot_save(
    manager: State<'_, ScreenshotManager>,
    image_base64: String,
    path: String,
) -> Result<String, String> {
    let image_data = base64::engine::general_purpose::STANDARD
        .decode(&image_base64)
        .map_err(|e| format!("Failed to decode image: {}", e))?;
    
    manager.save_to_file(&image_data, &path)
}

// ============== Screenshot History Commands ==============

/// Get screenshot history
#[tauri::command]
pub async fn screenshot_get_history(
    manager: State<'_, ScreenshotManager>,
    count: Option<usize>,
) -> Result<Vec<ScreenshotHistoryEntry>, String> {
    Ok(manager.get_history(count.unwrap_or(20)))
}

/// Search screenshot history by OCR text
#[tauri::command]
pub async fn screenshot_search_history(
    manager: State<'_, ScreenshotManager>,
    query: String,
) -> Result<Vec<ScreenshotHistoryEntry>, String> {
    Ok(manager.search_history_by_text(&query))
}

/// Get screenshot by ID
#[tauri::command]
pub async fn screenshot_get_by_id(
    manager: State<'_, ScreenshotManager>,
    id: String,
) -> Result<Option<ScreenshotHistoryEntry>, String> {
    Ok(manager.get_screenshot_by_id(&id))
}

/// Pin screenshot
#[tauri::command]
pub async fn screenshot_pin(
    manager: State<'_, ScreenshotManager>,
    id: String,
) -> Result<bool, String> {
    Ok(manager.pin_screenshot(&id))
}

/// Unpin screenshot
#[tauri::command]
pub async fn screenshot_unpin(
    manager: State<'_, ScreenshotManager>,
    id: String,
) -> Result<bool, String> {
    Ok(manager.unpin_screenshot(&id))
}

/// Delete screenshot from history
#[tauri::command]
pub async fn screenshot_delete(
    manager: State<'_, ScreenshotManager>,
    id: String,
) -> Result<bool, String> {
    Ok(manager.delete_screenshot(&id))
}

/// Clear screenshot history
#[tauri::command]
pub async fn screenshot_clear_history(
    manager: State<'_, ScreenshotManager>,
) -> Result<(), String> {
    manager.clear_history();
    Ok(())
}

// ============== Windows OCR Commands ==============

/// Extract text using Windows OCR
#[tauri::command]
pub async fn screenshot_ocr_windows(
    manager: State<'_, ScreenshotManager>,
    image_base64: String,
) -> Result<WinOcrResult, String> {
    let image_data = base64::engine::general_purpose::STANDARD
        .decode(&image_base64)
        .map_err(|e| format!("Failed to decode image: {}", e))?;
    
    manager.extract_text_windows(&image_data)
}

/// Get available OCR languages
#[tauri::command]
pub async fn screenshot_get_ocr_languages(
    manager: State<'_, ScreenshotManager>,
) -> Result<Vec<String>, String> {
    Ok(manager.get_ocr_languages())
}

// ============== Capture with History Commands ==============

/// Capture fullscreen and add to history
#[tauri::command]
pub async fn screenshot_capture_fullscreen_with_history(
    manager: State<'_, ScreenshotManager>,
    monitor_index: Option<usize>,
) -> Result<ScreenshotResult, String> {
    let result = manager.capture_fullscreen_with_history(monitor_index).await?;
    Ok(ScreenshotResult {
        image_base64: base64::engine::general_purpose::STANDARD.encode(&result.image_data),
        metadata: result.metadata,
    })
}

/// Capture window and add to history
#[tauri::command]
pub async fn screenshot_capture_window_with_history(
    manager: State<'_, ScreenshotManager>,
) -> Result<ScreenshotResult, String> {
    let result = manager.capture_window_with_history().await?;
    Ok(ScreenshotResult {
        image_base64: base64::engine::general_purpose::STANDARD.encode(&result.image_data),
        metadata: result.metadata,
    })
}

/// Capture region and add to history
#[tauri::command]
pub async fn screenshot_capture_region_with_history(
    manager: State<'_, ScreenshotManager>,
    x: i32,
    y: i32,
    width: u32,
    height: u32,
) -> Result<ScreenshotResult, String> {
    let region = CaptureRegion { x, y, width, height };
    let result = manager.capture_region_with_history(region).await?;
    Ok(ScreenshotResult {
        image_base64: base64::engine::general_purpose::STANDARD.encode(&result.image_data),
        metadata: result.metadata,
    })
}

/// Screenshot result for frontend
#[derive(serde::Serialize, serde::Deserialize)]
pub struct ScreenshotResult {
    pub image_base64: String,
    pub metadata: ScreenshotMetadata,
}

use base64::Engine;

// ============== Window Management Commands ==============

/// Get list of all visible windows
#[tauri::command]
pub async fn screenshot_get_windows(
    manager: State<'_, ScreenshotManager>,
) -> Result<Vec<WindowInfo>, String> {
    Ok(manager.get_windows())
}

/// Get list of all visible windows with thumbnails
#[tauri::command]
pub async fn screenshot_get_windows_with_thumbnails(
    manager: State<'_, ScreenshotManager>,
    thumbnail_size: Option<u32>,
) -> Result<Vec<WindowInfo>, String> {
    Ok(manager.get_windows_with_thumbnails(thumbnail_size.unwrap_or(160)))
}

/// Capture a specific window by its HWND
#[tauri::command]
pub async fn screenshot_capture_window_by_hwnd(
    manager: State<'_, ScreenshotManager>,
    hwnd: isize,
) -> Result<ScreenshotResult, String> {
    let result = manager.capture_window_by_hwnd(hwnd)?;
    Ok(ScreenshotResult {
        image_base64: base64::engine::general_purpose::STANDARD.encode(&result.image_data),
        metadata: result.metadata,
    })
}

/// Capture a specific window by HWND and add to history
#[tauri::command]
pub async fn screenshot_capture_window_by_hwnd_with_history(
    manager: State<'_, ScreenshotManager>,
    hwnd: isize,
) -> Result<ScreenshotResult, String> {
    let result = manager.capture_window_by_hwnd_with_history(hwnd).await?;
    Ok(ScreenshotResult {
        image_base64: base64::engine::general_purpose::STANDARD.encode(&result.image_data),
        metadata: result.metadata,
    })
}

/// Calculate snap position for window movement
#[tauri::command]
pub async fn screenshot_calculate_snap(
    manager: State<'_, ScreenshotManager>,
    window_hwnd: isize,
    proposed_x: i32,
    proposed_y: i32,
    window_width: u32,
    window_height: u32,
) -> Result<SnapResult, String> {
    Ok(manager.calculate_snap_position(
        window_hwnd,
        proposed_x,
        proposed_y,
        window_width,
        window_height,
    ))
}

/// Get current snap configuration
#[tauri::command]
pub async fn screenshot_get_snap_config(
    manager: State<'_, ScreenshotManager>,
) -> Result<SnapConfig, String> {
    Ok(manager.get_snap_config())
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::screenshot::{OcrLine, OcrWord, OcrBounds};

    #[test]
    fn test_screenshot_result_struct() {
        let result = ScreenshotResult {
            image_base64: "iVBORw0KGgoAAAANSUhEUgAAAAEAAAABCAYAAAAfFcSJAAAADUlEQVR42mNk+M9QDwADhgGAWjR9awAAAABJRU5ErkJggg==".to_string(),
            metadata: ScreenshotMetadata {
                timestamp: 1704067200000,
                width: 100,
                height: 100,
                mode: "fullscreen".to_string(),
                monitor_index: Some(0),
                window_title: None,
                region: None,
                file_path: None,
                ocr_text: None,
            },
        };
        
        assert!(!result.image_base64.is_empty());
        assert_eq!(result.metadata.width, 100);
        assert_eq!(result.metadata.height, 100);
    }

    #[test]
    fn test_screenshot_result_serialization() {
        let result = ScreenshotResult {
            image_base64: "base64data".to_string(),
            metadata: ScreenshotMetadata {
                timestamp: 1704067200000,
                width: 1920,
                height: 1080,
                mode: "fullscreen".to_string(),
                monitor_index: Some(0),
                window_title: None,
                region: None,
                file_path: Some("/path/to/screenshot.png".to_string()),
                ocr_text: None,
            },
        };
        
        let serialized = serde_json::to_string(&result).unwrap();
        assert!(serialized.contains("\"image_base64\":\"base64data\""));
        assert!(serialized.contains("\"width\":1920"));
        
        let deserialized: ScreenshotResult = serde_json::from_str(&serialized).unwrap();
        assert_eq!(result.image_base64, deserialized.image_base64);
        assert_eq!(result.metadata.width, deserialized.metadata.width);
    }

    #[test]
    fn test_capture_region_struct() {
        let region = CaptureRegion {
            x: 100,
            y: 200,
            width: 800,
            height: 600,
        };
        
        assert_eq!(region.x, 100);
        assert_eq!(region.y, 200);
        assert_eq!(region.width, 800);
        assert_eq!(region.height, 600);
    }

    #[test]
    fn test_capture_region_serialization() {
        let region = CaptureRegion {
            x: 0,
            y: 0,
            width: 1920,
            height: 1080,
        };
        
        let serialized = serde_json::to_string(&region).unwrap();
        let deserialized: CaptureRegion = serde_json::from_str(&serialized).unwrap();
        
        assert_eq!(region.x, deserialized.x);
        assert_eq!(region.y, deserialized.y);
        assert_eq!(region.width, deserialized.width);
        assert_eq!(region.height, deserialized.height);
    }

    #[test]
    fn test_capture_region_negative_coords() {
        let region = CaptureRegion {
            x: -100,
            y: -50,
            width: 500,
            height: 300,
        };
        
        assert_eq!(region.x, -100);
        assert_eq!(region.y, -50);
    }

    #[test]
    fn test_monitor_info_struct() {
        let monitor = MonitorInfo {
            index: 0,
            name: "Primary Monitor".to_string(),
            x: 0,
            y: 0,
            width: 1920,
            height: 1080,
            is_primary: true,
            scale_factor: 1.0,
        };
        
        assert_eq!(monitor.index, 0);
        assert!(monitor.is_primary);
        assert_eq!(monitor.scale_factor, 1.0);
    }

    #[test]
    fn test_monitor_info_high_dpi() {
        let monitor = MonitorInfo {
            index: 1,
            name: "Retina Display".to_string(),
            x: 1920,
            y: 0,
            width: 2560,
            height: 1440,
            is_primary: false,
            scale_factor: 2.0,
        };
        
        assert_eq!(monitor.scale_factor, 2.0);
        assert!(!monitor.is_primary);
    }

    #[test]
    fn test_screenshot_config_default() {
        let config = ScreenshotConfig::default();
        
        assert_eq!(config.format, "png");
        assert_eq!(config.quality, 95);
        assert!(!config.include_cursor);
    }

    #[test]
    fn test_screenshot_config_custom() {
        let config = ScreenshotConfig {
            save_directory: Some("/custom/path".to_string()),
            format: "jpg".to_string(),
            quality: 80,
            include_cursor: true,
            delay_ms: 500,
            copy_to_clipboard: false,
            show_notification: false,
            ocr_language: "chi_sim".to_string(),
        };
        
        assert_eq!(config.save_directory, Some("/custom/path".to_string()));
        assert_eq!(config.format, "jpg");
        assert_eq!(config.quality, 80);
        assert!(config.include_cursor);
    }

    #[test]
    fn test_screenshot_config_struct() {
        let config = ScreenshotConfig {
            save_directory: Some("/screenshots".to_string()),
            format: "png".to_string(),
            quality: 90,
            include_cursor: false,
            delay_ms: 0,
            copy_to_clipboard: true,
            show_notification: true,
            ocr_language: "en".to_string(),
        };
        
        assert_eq!(config.format, "png");
        assert_eq!(config.quality, 90);
        assert!(!config.include_cursor);
        assert!(config.copy_to_clipboard);
    }

    #[test]
    fn test_screenshot_config_serialization() {
        let config = ScreenshotConfig {
            save_directory: None,
            format: "jpeg".to_string(),
            quality: 80,
            include_cursor: true,
            delay_ms: 500,
            copy_to_clipboard: false,
            show_notification: false,
            ocr_language: "zh-CN".to_string(),
        };
        
        let serialized = serde_json::to_string(&config).unwrap();
        let deserialized: ScreenshotConfig = serde_json::from_str(&serialized).unwrap();
        
        assert_eq!(config.format, deserialized.format);
        assert_eq!(config.quality, deserialized.quality);
        assert_eq!(config.ocr_language, deserialized.ocr_language);
    }

    #[test]
    fn test_screenshot_metadata_struct() {
        let metadata = ScreenshotMetadata {
            timestamp: 1704067200000,
            width: 1920,
            height: 1080,
            mode: "window".to_string(),
            monitor_index: None,
            window_title: Some("Test Window".to_string()),
            region: None,
            file_path: None,
            ocr_text: None,
        };
        
        assert_eq!(metadata.width, 1920);
        assert_eq!(metadata.mode, "window");
        assert_eq!(metadata.window_title, Some("Test Window".to_string()));
    }

    #[test]
    fn test_screenshot_history_entry_struct() {
        let entry = ScreenshotHistoryEntry::new(800, 600, "region")
            .with_thumbnail("thumb_data".to_string())
            .with_ocr_text("Extracted text".to_string());
        
        assert!(!entry.id.is_empty());
        assert!(!entry.is_pinned);
        assert_eq!(entry.width, 800);
        assert_eq!(entry.height, 600);
        assert_eq!(entry.ocr_text, Some("Extracted text".to_string()));
    }

    #[test]
    fn test_win_ocr_result_struct() {
        let result = WinOcrResult {
            text: "Hello World".to_string(),
            confidence: 0.95,
            lines: vec![
                OcrLine {
                    text: "Hello World".to_string(),
                    words: vec![
                        OcrWord {
                            text: "Hello".to_string(),
                            bounds: OcrBounds { x: 0.0, y: 0.0, width: 50.0, height: 20.0 },
                            confidence: 0.95,
                        },
                        OcrWord {
                            text: "World".to_string(),
                            bounds: OcrBounds { x: 55.0, y: 0.0, width: 50.0, height: 20.0 },
                            confidence: 0.95,
                        },
                    ],
                    bounds: OcrBounds { x: 0.0, y: 0.0, width: 105.0, height: 20.0 },
                },
            ],
            language: Some("en".to_string()),
        };
        
        assert_eq!(result.text, "Hello World");
        assert_eq!(result.confidence, 0.95);
        assert_eq!(result.lines.len(), 1);
    }

    #[test]
    fn test_win_ocr_result_empty() {
        let result = WinOcrResult {
            text: String::new(),
            confidence: 0.0,
            lines: vec![],
            language: None,
        };
        
        assert!(result.text.is_empty());
        assert!(result.lines.is_empty());
        assert!(result.language.is_none());
    }

    #[test]
    fn test_ocr_bounds_struct() {
        let bounds = OcrBounds {
            x: 10.0,
            y: 20.0,
            width: 100.0,
            height: 30.0,
        };
        
        assert_eq!(bounds.x, 10.0);
        assert_eq!(bounds.y, 20.0);
        assert_eq!(bounds.width, 100.0);
        assert_eq!(bounds.height, 30.0);
    }
}

File: lib/native/screenshot.ts
/**
 * Screenshot Native API
 *
 * Provides TypeScript bindings for the Tauri screenshot commands.
 */

import { invoke } from "@tauri-apps/api/core";

// ============== Types ==============

export interface ScreenshotMetadata {
  width: number;
  height: number;
  mode: string;
  timestamp: number;
  window_title?: string;
  monitor_index?: number;
}

export interface ScreenshotResult {
  image_base64: string;
  metadata: ScreenshotMetadata;
}

export interface CaptureRegion {
  x: number;
  y: number;
  width: number;
  height: number;
}

export interface MonitorInfo {
  index: number;
  name: string;
  x: number;
  y: number;
  width: number;
  height: number;
  is_primary: boolean;
  scale_factor: number;
}

export interface ScreenshotConfig {
  save_directory?: string;
  format: string;
  quality: number;
  include_cursor: boolean;
  delay_ms?: number;
  copy_to_clipboard: boolean;
  show_notification: boolean;
  ocr_language?: string;
  auto_save: boolean;
  filename_template: string;
}

export interface ScreenshotHistoryEntry {
  id: string;
  timestamp: number;
  thumbnail_base64?: string;
  file_path?: string;
  width: number;
  height: number;
  mode: string;
  window_title?: string;
  ocr_text?: string;
  label?: string;
  tags: string[];
  is_pinned: boolean;
}

export interface WinOcrResult {
  text: string;
  lines: OcrLine[];
  language?: string;
  confidence: number;
}

export interface OcrLine {
  text: string;
  words: OcrWord[];
  bounds: OcrBounds;
}

export interface OcrWord {
  text: string;
  bounds: OcrBounds;
  confidence: number;
}

export interface OcrBounds {
  x: number;
  y: number;
  width: number;
  height: number;
}

export type Annotation =
  | {
      type: 'Rectangle';
      x: number;
      y: number;
      width: number;
      height: number;
      color: string;
      stroke_width: number;
      filled: boolean;
    }
  | {
      type: 'Ellipse';
      cx: number;
      cy: number;
      rx: number;
      ry: number;
      color: string;
      stroke_width: number;
      filled: boolean;
    }
  | {
      type: 'Arrow';
      start_x: number;
      start_y: number;
      end_x: number;
      end_y: number;
      color: string;
      stroke_width: number;
    }
  | {
      type: 'Freehand';
      points: Array<[number, number]>;
      color: string;
      stroke_width: number;
    }
  | {
      type: 'Text';
      x: number;
      y: number;
      text: string;
      font_size: number;
      color: string;
      background?: string | null;
    }
  | {
      type: 'Blur';
      x: number;
      y: number;
      width: number;
      height: number;
      intensity: number;
    }
  | {
      type: 'Highlight';
      x: number;
      y: number;
      width: number;
      height: number;
      color: string;
      opacity: number;
    }
  | {
      type: 'Marker';
      x: number;
      y: number;
      number: number;
      color: string;
      size: number;
    };

export interface AnnotatedScreenshotResult {
  image_base64: string;
}

export interface SelectionValidationResult {
  region: CaptureRegion;
  is_valid: boolean;
}

export interface SnapConfig {
  snap_distance: number;
  snap_to_screen: boolean;
  snap_to_windows: boolean;
  snap_to_elements: boolean;
  show_guide_lines: boolean;
  magnetic_edges: boolean;
}

export interface WindowInfo {
  hwnd: number;
  title: string;
  process_name: string;
  pid: number;
  x: number;
  y: number;
  width: number;
  height: number;
  is_minimized: boolean;
  is_maximized: boolean;
  is_visible: boolean;
  thumbnail_base64?: string;
}

export interface ElementInfo {
  x: number;
  y: number;
  width: number;
  height: number;
  element_type: string;
  name?: string;
  parent_hwnd: number;
}

export interface SnapGuide {
  orientation: 'horizontal' | 'vertical';
  position: number;
  start: number;
  end: number;
  source: string;
}

export interface SelectionSnapResult {
  x: number;
  y: number;
  width: number;
  height: number;
  snapped: boolean;
  guides: SnapGuide[];
}

// ============== Basic Capture Functions ==============

/**
 * Capture full screen screenshot
 */
export async function captureFullscreen(
  monitorIndex?: number
): Promise<ScreenshotResult> {
  return invoke("screenshot_capture_fullscreen", { monitorIndex });
}

/**
 * Capture active window screenshot
 */
export async function captureWindow(): Promise<ScreenshotResult> {
  return invoke("screenshot_capture_window");
}

/**
 * Capture a specific region
 */
export async function captureRegion(
  x: number,
  y: number,
  width: number,
  height: number
): Promise<ScreenshotResult> {
  return invoke("screenshot_capture_region", { x, y, width, height });
}

/**
 * Start interactive region selection
 */
export async function startRegionSelection(): Promise<CaptureRegion> {
  return invoke("screenshot_start_region_selection");
}

// ============== Capture with History ==============

/**
 * Capture fullscreen and add to history
 */
export async function captureFullscreenWithHistory(
  monitorIndex?: number
): Promise<ScreenshotResult> {
  return invoke("screenshot_capture_fullscreen_with_history", { monitorIndex });
}

/**
 * Capture window and add to history
 */
export async function captureWindowWithHistory(): Promise<ScreenshotResult> {
  return invoke("screenshot_capture_window_with_history");
}

/**
 * Capture region and add to history
 */
export async function captureRegionWithHistory(
  x: number,
  y: number,
  width: number,
  height: number
): Promise<ScreenshotResult> {
  return invoke("screenshot_capture_region_with_history", { x, y, width, height });
}

// ============== Annotation Functions ==============

/**
 * Apply annotations to screenshot image data
 */
export async function applyAnnotations(
  imageBase64: string,
  annotations: Annotation[]
): Promise<AnnotatedScreenshotResult> {
  return invoke("screenshot_apply_annotations", { imageBase64, annotations });
}

// ============== Selection Functions ==============

/**
 * Validate selection state and normalize region
 */
export async function validateSelection(
  startX: number,
  startY: number,
  currentX: number,
  currentY: number
): Promise<SelectionValidationResult> {
  return invoke("screenshot_validate_selection", {
    startX,
    startY,
    currentX,
    currentY,
  });
}

// ============== OCR Functions ==============

/**
 * Extract text from screenshot using basic OCR
 */
export async function extractText(imageBase64: string): Promise<string> {
  return invoke("screenshot_ocr", { imageBase64 });
}

/**
 * Extract text using Windows OCR (more accurate)
 */
export async function extractTextWindows(
  imageBase64: string
): Promise<WinOcrResult> {
  return invoke("screenshot_ocr_windows", { imageBase64 });
}

/**
 * Extract text using Windows OCR with specified language
 */
export async function extractTextWithLanguage(
  imageBase64: string,
  language?: string
): Promise<WinOcrResult> {
  return invoke("screenshot_ocr_with_language", { imageBase64, language });
}

/**
 * Get available OCR languages installed on the system
 */
export async function getOcrLanguages(): Promise<string[]> {
  return invoke("screenshot_get_ocr_languages");
}

/**
 * Check if OCR is available on this system
 */
export async function isOcrAvailable(): Promise<boolean> {
  return invoke("screenshot_ocr_is_available");
}

/**
 * Check if a specific OCR language is available
 */
export async function isOcrLanguageAvailable(language: string): Promise<boolean> {
  return invoke("screenshot_ocr_is_language_available", { language });
}

/**
 * Set Windows OCR language used by screenshot manager
 */
export async function setOcrLanguage(language: string): Promise<void> {
  return invoke("screenshot_set_ocr_language", { language });
}

// ============== History Functions ==============

/**
 * Get screenshot history
 */
export async function getHistory(count?: number): Promise<ScreenshotHistoryEntry[]> {
  return invoke("screenshot_get_history", { count });
}

/**
 * Search screenshot history by OCR text
 */
export async function searchHistory(query: string): Promise<ScreenshotHistoryEntry[]> {
  return invoke("screenshot_search_history", { query });
}

/**
 * Search screenshot history by label
 */
export async function searchHistoryByLabel(
  label: string
): Promise<ScreenshotHistoryEntry[]> {
  return invoke("screenshot_search_history_by_label", { label });
}

/**
 * Get all screenshot history entries
 */
export async function getAllHistory(): Promise<ScreenshotHistoryEntry[]> {
  return invoke("screenshot_get_all_history");
}

/**
 * Get pinned screenshot history
 */
export async function getPinnedHistory(): Promise<ScreenshotHistoryEntry[]> {
  return invoke("screenshot_get_pinned_history");
}

/**
 * Get screenshot by ID
 */
export async function getScreenshotById(
  id: string
): Promise<ScreenshotHistoryEntry | null> {
  return invoke("screenshot_get_by_id", { id });
}

/**
 * Pin a screenshot
 */
export async function pinScreenshot(id: string): Promise<boolean> {
  return invoke("screenshot_pin", { id });
}

/**
 * Unpin a screenshot
 */
export async function unpinScreenshot(id: string): Promise<boolean> {
  return invoke("screenshot_unpin", { id });
}

/**
 * Delete a screenshot from history
 */
export async function deleteScreenshot(id: string): Promise<boolean> {
  return invoke("screenshot_delete", { id });
}

/**
 * Clear screenshot history
 */
export async function clearHistory(): Promise<void> {
  return invoke("screenshot_clear_history");
}

/**
 * Clear all screenshot history
 */
export async function clearAllHistory(): Promise<void> {
  return invoke("screenshot_clear_all_history");
}

/**
 * Get history stats (count, is_empty)
 */
export async function getHistoryStats(): Promise<[number, boolean]> {
  return invoke("screenshot_get_history_stats");
}

// ============== Configuration ==============

/**
 * Get screenshot configuration
 */
export async function getConfig(): Promise<ScreenshotConfig> {
  return invoke("screenshot_get_config");
}

/**
 * Update screenshot configuration
 */
export async function updateConfig(config: ScreenshotConfig): Promise<void> {
  return invoke("screenshot_update_config", { config });
}

/**
 * Get available monitors
 */
export async function getMonitors(): Promise<MonitorInfo[]> {
  return invoke("screenshot_get_monitors");
}

/**
 * Get snap configuration
 */
export async function getSnapConfig(): Promise<SnapConfig> {
  return invoke("screenshot_get_snap_config");
}

/**
 * Set snap configuration
 */
export async function setSnapConfig(config: SnapConfig): Promise<void> {
  return invoke("screenshot_set_snap_config", { config });
}

/**
 * Save screenshot to file
 */
export async function saveToFile(
  imageBase64: string,
  path: string
): Promise<string> {
  return invoke("screenshot_save", { imageBase64, path });
}

// ============== Window Detection Functions ==============

/**
 * Get all visible windows
 */
export async function getWindows(): Promise<WindowInfo[]> {
  return invoke("screenshot_get_windows");
}

/**
 * Get all visible windows with thumbnails
 */
export async function getWindowsWithThumbnails(
  thumbnailSize?: number
): Promise<WindowInfo[]> {
  return invoke("screenshot_get_windows_with_thumbnails", { thumbnailSize });
}

/**
 * Get the window at a specific screen point (for auto-detection during selection)
 */
export async function getWindowAtPoint(
  x: number,
  y: number
): Promise<WindowInfo | null> {
  return invoke("screenshot_get_window_at_point", { x, y });
}

/**
 * Get child elements of a window (for element-level detection)
 */
export async function getChildElements(
  hwnd: number,
  maxDepth?: number
): Promise<ElementInfo[]> {
  return invoke("screenshot_get_child_elements", { hwnd, maxDepth });
}

/**
 * Capture a specific window by its HWND
 */
export async function captureWindowByHwnd(hwnd: number): Promise<ScreenshotResult> {
  return invoke("screenshot_capture_window_by_hwnd", { hwnd });
}

/**
 * Capture a specific window by HWND and add to history
 */
export async function captureWindowByHwndWithHistory(
  hwnd: number
): Promise<ScreenshotResult> {
  return invoke("screenshot_capture_window_by_hwnd_with_history", { hwnd });
}

// ============== Selection Snap Functions ==============

/**
 * Calculate snapped selection rectangle during region selection
 * Returns adjusted coordinates if edges are near window/screen boundaries
 */
export async function calculateSelectionSnap(
  selectionX: number,
  selectionY: number,
  selectionWidth: number,
  selectionHeight: number
): Promise<SelectionSnapResult> {
  return invoke("screenshot_calculate_selection_snap", {
    selectionX,
    selectionY,
    selectionWidth,
    selectionHeight,
  });
}

// ============== Color Picker Functions ==============

/**
 * Get pixel color at screen coordinates
 * Returns hex color string (e.g., "#FF5733") or null if failed
 */
export async function getPixelColor(x: number, y: number): Promise<string | null> {
  return invoke("screenshot_get_pixel_color", { x, y });
}
